Xamarin.Forms.Controls.Issues/Xamarin.Forms.Controls.Issues.Shared/Github5623.xaml.cs

```diff
@@ -0,0 +1,201 @@
+	public partial class Github5623 : TestContentPage
+	{
+		int _itemCount = 10;
+		int _maximumItemCount = 100;
+		int _pageSize = 10;
+		static SemaphoreSlim SemaphoreSlim = new SemaphoreSlim(1, 1);
+
+		public Github5623()
+		{
+#if APP
+			Device.SetFlags(new List<string> { CollectionView.CollectionViewExperimental });
+
+			InitializeComponent();
+
+			BindingContext = new ViewModel5623();
+#endif
+		}
+
+		protected override void Init()
+		{
+
+		}
+
+		async void CollectionView_RemainingItemsThresholdReached(object sender, System.EventArgs e)
+		{
+			await SemaphoreSlim.WaitAsync();
+			try
+			{
+				var itemsSource = (sender as CollectionView).ItemsSource as ObservableCollection<Model5623>;
+				var nextSet = await GetNextSetAsync();
+
+				// nothing to add
+				if (nextSet.Count == 0)
+					return;
+
+				Device.BeginInvokeOnMainThread(() =>
+				{
+					foreach (var item in nextSet)
+					{
+						itemsSource.Add(item);
+					}
+				});
+
+				System.Diagnostics.Debug.WriteLine("Count: " + itemsSource.Count);
+			}
+			finally
+			{
+				SemaphoreSlim.Release();
+			}
+		}
+
+		void CollectionView_OnScrolled(object sender, ItemsViewScrolledEventArgs e)
+		{
+			Label1.Text = "HorizontalDelta: " + e.HorizontalDelta;
+			Label2.Text = "VerticalDelta: " + e.VerticalDelta;
+			Label3.Text = "HorizontalOffset: " + e.HorizontalOffset;
+			Label4.Text = "VerticalOffset: " + e.VerticalOffset;
+			Label5.Text = "FirstVisibleItemIndex: " + e.FirstVisibleItemIndex;
+			Label6.Text = "CenterItemIndex: " + e.CenterItemIndex;
+			Label7.Text = "LastVisibleItemIndex: " + e.LastVisibleItemIndex;
+		}
+
+		async Task<ObservableCollection<Model5623>> GetNextSetAsync()
+		{
+			return await Task.Run(() =>
+			{
+				var collection = new ObservableCollection<Model5623>();
+				var count = _pageSize;
+
+				if (_itemCount + count > _maximumItemCount)
+					count = _maximumItemCount - _itemCount;
+
+				for (var i = _itemCount; i < _itemCount + count; i++)
+				{
+					collection.Add(new Model5623((BindingContext as ViewModel5623).ItemSizingStrategy == ItemSizingStrategy.MeasureAllItems)
+					{
+						Text = i.ToString(),
+						BackgroundColor = i % 2 == 0 ? Color.AntiqueWhite : Color.Lavender
+					});
+				}
+
+				_itemCount += count;
+
+				return collection;
+			});
+		}
+
+#if UITEST
+		[Test]
+		public void CollectionViewInfiniteScroll()
+		{
+			RunningApp.WaitForElement (q => q.Marked ("CollectionView"));
+			
+			var elementQuery = RunningApp.Query(c => c.Marked((_maximumItemCount - 1).ToString()));
+			var elementAvailable = elementQuery.Any();
+			var attempt = 0;
+
+			while (!elementAvailable)
+			{
+				RunningApp.ScrollDown();
+
+				elementAvailable = elementQuery.Any();
+				if (elementAvailable)
+					break;
+
+				++attempt;
+
+				if (attempt == 30)
+				{
+					Assert.Fail("Failed to find the last element");
```

Comment:
After adding the above, it then fails on this line. It seems the UI test needs some work :)

Reply:
As I've told the team before, I haven't been able to run the UI tests due to errors with my setup. In the past, I've set the `ANDROID_HOME` variable, but the test library wasn't able to find it. I've also seen dlls constantly locked on the C drive along with symbols not loaded. It'd be great if Xamarin can provide clear instructions on how to run the UI tests in this solution.

Reply:
My apologies, I must have overlooked that. I'm assuming you're on Windows then, since you're not mentioning iOS?

Reply:
That said, if the UI test continues to fail, I'll let the team take it from here and fix it for us all. :)

Reply:
Correct. Win 10.

Reply:
Are you on the Xamarin Chat slack? Or have some other way to reach you? Or worst case can you reach out to gerald.versluis@microsoft.com? Let's see if we can get it to work and from there I'll create a little writeup to make sure this doesn't happen to other people

Reply:
Emailed.